### DISCUSSION TITLE
Introspection

### DISCUSSION THREAD

**Discussion by deltamarnix** [pyphoenix/pyphoenix-project#50] [Ideas]
DFN information is currently written by the class generator as a class variable in every package:

```py
dfn = [
        ["header", "multi-package", "package-type stress-package"],
        [
            "block options",
            "name readasarrays",
            "type keyword",
            "shape",
            "reader urword",
            "optional false",
            "default_value True",
        ],
        [
            "block options",
            "name fixed_cell",
            "type keyword",
            "shape",
            "reader urword",
            "optional true",
        ],
        ...
]
```

Reading this information during runtime is hard. Conversion to toml already made it easier to understand the nested structure of this information. It should be possible for a user to inspect the dictionaray that comes out of the toml.

Besides that there is also the `MFStructure` class, which is a singleton class accessed by models and packages to obtain information on the definitions of the different packages.

It is likely that we only need one definition of what the package contains and I expect that the [`attrs` package](https://www.attrs.org/en/stable/api.html#attrs.fields) will give us and the user enough information for introspection.

**Comment by wpbonelli** [comment]
As a near-term flopy3 refactoring step after https://github.com/modflowpy/flopy/pull/2333, @mjreno and I discussed replacing the list of lists with the dictionary loaded from TOML, in hopes of dramatically simplifying `mfstructure.py`.

Totally agreed on `attrs` for 4.x.
